torch/fx/passes/runtime_assert.py

```diff
@@ -0,0 +1,241 @@
+                # https://twitter.com/ezyang/status/1745801370299482492
+                # We actually made an attempt at this in
+                # https://github.com/pytorch/pytorch/pull/119043
+                # which didn't work.
+                #
+                # Another ideas for how to do this:
+                # - Have bound_sympy be the source of truth of the ranges of any expression
+                # - Cache intermediate results for every subexpression of bound_sympy
+                # - This cache should be possible to edit to refine ranges
+                #
+                # One issue with this proposal is that if
+                # we have a bound on 2x, we are not going to be able to
+                # apply it for 4x.  Similarly, we may have bounds for an
+                # equivalent expression that we are not applying because
+                # it's not a perfect match (e.g. x < y vs y > x)".
+                #
+                # The first issue we already have it and it's impossible
+                # to solve in general, so any implementation on a best
+                # effort basis should do.
+                #
+                # The second issue is a preexisting one. It can be mitigated
+                # with a normalisation algorithm. In general, it may also
+                # be on a best effort basis, but since our grammar is not
+                # terribly difficult, chances are we could even fully
+                # normalise SymPy expressions... who knows.
+
+                if i0 in shape_env.size_like:
+                    graph.call_function(
+                        torch._check_is_size, (symbol_to_proxy[i0].node,)
+                    )
+
+                vr = shape_env.var_to_range[i0]
+                if not shape_env._default_unspecified_value_range().issubset(vr):
+                    # The runtime range is constrained, so add a runtime
+                    # assert and also explicitly refine the range
+                    # (refinement should not be necessary once runtime
+                    # asserts cause refinement, but that's NYI)
+                    def convert(s):
+                        try:
+                            return int(s)
+                        except TypeError:
+                            return None
+
+                    ge_check = graph.call_function(
+                        operator.ge,
+                        (
+                            symbol_to_proxy[i0].node,
+                            convert(vr.lower),
+                        ),
+                    )
+                    le_check = graph.call_function(
+                        operator.le,
+                        (
+                            symbol_to_proxy[i0].node,
+                            convert(vr.upper),
+                        ),
+                    )
+                    graph.call_function(torch._check, (ge_check,))
+                    graph.call_function(torch._check, (le_check,))
```

Comment:
Should these be `torch.ops.aten._async_assert.msg(ge_check, "msg")`? Because if we're running the pass after tracing non-strict, then all ops inserted need to be OpOverloads